stan/math/opencl/matrix_cl.hpp

```diff
@@ -236,18 +235,20 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
    * @throw <code>std::invalid_argument</code> if the
    * matrices do not have matching dimensions
    */
-  template <int R, int C>
-  explicit matrix_cl(const std::vector<Eigen::Matrix<T, R, C>>& A) try
-      : rows_(A.empty() ? 0 : A[0].size()),
-        cols_(A.size()) {
+  template <typename Vec, require_std_vector_vt<is_eigen, Vec>...,
+            require_std_vector_st<std::is_arithmetic, Vec>...>
+  explicit matrix_cl(Vec&& A) try : rows_(A.empty() ? 0 : A[0].size()),
+                                    cols_(A.size()) {
     if (this->size() == 0) {
       return;
     }
+
     cl::Context& ctx = opencl_context.context();
     cl::CommandQueue& queue = opencl_context.queue();
     // creates the OpenCL buffer to copy the Eigen
     // matrix to the OpenCL device
     buffer_cl_ = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(T) * size());
+    cl::Event write_event;
```

Comment:
Why did you move this outside the loop? We need to add all write events, not just the last one.

Reply:
Ah yes! So, do we? I think it depends on whether OpenCL does async when it writes to an individual buffer or if writes to a single buffer happen sequentially. Maybe there is not a guarantee around that. If I can find something in the docs that shows that is true then I'll revert this and else we can discuss it.

Reply:
Yeah I think we do, fixed this